jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpChannelOverHTTP2.java

```diff
@@ -261,31 +264,265 @@ public InvocationType getInvocationType()
             handle |= handleContent | handleRequest;
         }
 
+        boolean woken = getRequest().getHttpInput().wakeup();
         if (LOG.isDebugEnabled())
         {
-            LOG.debug("HTTP2 Request #{}/{}: {} bytes of {} content, handle: {}",
+            Stream stream = getStream();
+            LOG.debug("HTTP2 Request #{}/{}: {} bytes of {} content, woken: {}, handle: {}",
                     stream.getId(),
                     Integer.toHexString(stream.getSession().hashCode()),
                     length,
                     endStream ? "last" : "some",
+                    woken,
                     handle);
         }
+        handle |= woken;
 
         boolean wasDelayed = _delayedUntilContent;
         _delayedUntilContent = false;
         return handle || wasDelayed ? this : null;
     }
 
+    private static final HttpInput.Content EOF = new HttpInput.EofContent();
+    private static final HttpInput.Content DEMANDING = new HttpInput.SpecialContent() {};
+
+    private class ContentDemander
+    {
+        private final AtomicReference<HttpInput.Content> _content = new AtomicReference<>();
+
+        public void recycle()
+        {
+            HttpInput.Content c = _content.getAndSet(null);
+            if (c != null && !c.isSpecial())
+                throw new AssertionError("unconsumed content: " + c);
+        }
+
+        public HttpInput.Content take()
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c == null || c.isSpecial() || _content.compareAndSet(c, c.isEof() ? EOF : null))
+                    return c;
+            }
+        }
+
+        public boolean demand()
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c == DEMANDING)
+                    return false;
+                if (c != null)
+                    return true;
+                if (_content.compareAndSet(null, DEMANDING))
+                {
+                    IStream stream = getStream();
+                    if (stream == null)
+                    {
+                        _content.set(null);
+                        return false;
+                    }
+                    stream.demand(1);
+                    c = _content.get();
+                    return c != DEMANDING && c != null;
+                }
+            }
+        }
+
+        public void onContent(HttpInput.Content content)
+        {
+            while (true)
+            {
+                HttpInput.Content c = _content.get();
+                if (c == DEMANDING)
+                {
+                    if (_content.compareAndSet(DEMANDING, content))
+                        return;
+                }
+                else if (c == null)
+                {
+                    if (!content.isSpecial())
+                        throw new AssertionError("Non special content without demand");
```

Comment:
Well, if we reach this branch, that means the state machine is broken and we have a bug. I could fail the content before throwing but there's no guarantee that this will help.